docs/detailed_design/config-reference.md
Config reference
================

For each field the following notation is used

- (==1?) means the field is OPTIONAL and ONLY ONE can exist
- (>=1?) means the field is OPTIONAL and MORE THAN ONE can exist
- (==1!) means the field is REQUIRED and ONLY ONE can exist
- (>=1!) means the field is REQUIRED and MORE THAN ONE can exist

`.porter/config`

- [service_name](#service_name) (==1!)
- [porter_version](#porter_version) (==1!)
- [environments](#environments) (>=1!)
  - [name](#environment-name) (>=1!)
  - [stack_definition_path](#stack_definition_path) (==1?)
  - [autowire_security_groups](#autowire_security_groups) (==1?)
  - [role_arn](#role_arn) (==1!)
  - [instance_count](#instance_count) (==1?)
  - [instance_type](#instance_type) (==1?)
  - [blackout_windows](#blackout_windows) (>=1?)
  - [hot_swap](#hot_swap) (==1?)
  - [haproxy](==1?)
    - [request_header_captures](#header-captures) (>=1?)
    - [response_header_captures](#header-captures) (>=1?)
    - [log](#log) (==1?)
    - [compression](#compression) (==1?)
    - [compress_types](#compress_types) (==1?)
    - [timeout](#timeout)
    - [maxconn](#maxconn) (==1?)
    - [ssl](#ssl) (==1?)
      - [cert_directory](#cert_directory) (==1?)
      - [pem](#pem) (==?!)
      - [https_only](#https_only) (==??)
      - [https_redirect](#https_redirect) (==??)
  - [regions](#regions) (>=1!)
    - [name](#region-name) (==1!)
    - [stack_definition_path](#stack_definition_path) (==1?)
    - [vpc_id](#vpc_id) (==1?)
    - [role_arn](#role_arn) (==1!)
    - [ssl_cert_arn](#ssl_cert_arn) (==1?)
    - [hosted_zone_name](#hosted_zone_name) (==1?)
    - [instance_count](#instance_count) (==1?)
    - auto_scaling_group
      - [security_group_egress](#security_group_egress) (==1?)
      - [secrets_exec_name](#secrets_exec_name) (==1?)
      - [secrets_exec_args](#secrets_exec_args) (==1?)
    - [key_pair_name](#key_pair_name) (==1?)
    - [s3_bucket](#s3_bucket) (==1!)
    - [sse_kms_key_id](#sse_kms_key_id) (==1!)
    - [elb](#elb) (==1?)
    - [azs](#azs) (>=1!)
      - name
      - [subnet_id](#subnet_id) (==1?)
    - [containers](#containers) (>=1?)
      - name
      - [topology](#topology) (==1?)
      - [inet_port](#inet_port) (==1?)
      - [dockerfile](#container-dockerfile) (==1?)
      - [dockerfile_build](#container-dockerfile-build) (==1?)
      - [uid](#uid) (==1?)
      - [read_only](#read_only) (==1?)
      - [health_check](#health_check) (==1?)
      - [src_env_file](#src_env_file) (==1?)
      - [pids_limit](#pids_limit) (==1?)
- [hooks](#hooks) (==1?)
  - pre_pack (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - post_pack (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - pre_provision (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - post_provision (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - pre_promote (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - post_promote (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - pre_prune (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - post_prune (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - ec2_bootstrap (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)
  - [user_defined](#user-defined-hooks) (==1?)
    - [repo](#repo) (==1!)
    - [ref](#ref) (==1!)
    - [dockerfile](#hook-dockerfile) (==1?)
    - [environment](#hook-environment) (==1?)
    - [concurrent](#concurrent) (==1?)
    - [run_condition](#run_condition) (==1?)

### service_name

service_name is used to form CloudFormation stack names as well as "folders" in
the bucket that a service is uploaded into.

Must match `/^[a-zA-Z][-a-zA-Z0-9]*$/`

### porter_version

porter_version is used by deployment servers to download the correct version of
porter before running porter build commands

Must match `/^v\d+\.\d+\.\d+$/`

### environments

environments is a namespace for configuration

### environment name

Must match `/^[0-9a-zA-Z]+$/`

### stack_definition_path

stack_definition_path is a relative path from the `.porter/config` to a
CloudFormation stack definition file.

The most specific definition is used meaning if it's defined on an environment
and an environment's region, the region value will be used.

### autowire_security_groups

By default porter manages security groups to allow the provisioned ELB, and
inspects the ELB that instances will be promoted into so that both ELBs can send
traffic to EC2 instances.

Set `autowire_security_groups: false` to disable this.

This setting does not affect, and is not affected by, `security_group_egress`

### role_arn

role_arn is the IAM Role that porter will call AssumeRole on in order to perform
the AWS API actions needed to create, update, rollback, and delete
CloudFormation stacks and the resources they define.

It is marked optional but is required either on the environment or region. If
both are specified the region value will be used.

Example ARN

```
arn:aws:iam::123456789012:role/porter-deployment
```

### instance_count

instance_count is the desired number of instance per environment-region.

The default is 1.

### instance_type

instance_type the EC2 instance type to be used.

The default is m3.medium

Acceptable values are

```
c1.medium
c1.xlarge
c3.2xlarge
c3.4xlarge
c3.8xlarge
c3.large
c3.xlarge
cc2.8xlarge
hs1.8xlarge
i1.4xlarge
i2.2xlarge
i2.4xlarge
i2.8xlarge
i2.xlarge
m1.large
m1.medium
m1.small
m1.xlarge
m2.2xlarge
m2.4xlarge
m2.xlarge
m3.2xlarge
m3.large
m3.medium
m3.xlarge
```

### blackout_windows

blackout_window contains a start_time and end_time between which
'porter build ...' commands will exit with status 1 immediately.

The times are parsed by https://golang.org/pkg/time/#Parse with layout RFC3339
from https://golang.org/pkg/time/#pkg-constants

The blackout window is considered active if

1. The current system time is greater than start_time and less than end_time
1. time.Parse returns an error when parsing start_time or end_time
1. start_time > end_time

A sample blackout_window

```
blackout_windows:
- start_time: 2015-01-02T15:04:05Z07:00
  end_time: 2015-01-03T15:04:05Z07:00
```

### hot_swap

Opt into [hot swap deployments](hotswap.md)

```yaml
environments:
- name: stage
  hot_swap: true
```

### header-captures

Header captures can be defined. See the [HAProxy docs](https://cbonte.github.io/haproxy-dconv/1.5/configuration.html#8.8)
for more about how to parse these logs.

```yaml
environments:
- name: dev
  haproxy:
    request_header_captures:
    - header: X-Request-Id
      length: 40
    - header: X-Forwarded-For
      length: 45

    response_header_captures:
    - header: X-Request-Id
      length: 40
```

### log

Turn off HAProxy logs which saves CPU cycles

```yaml
environments:
- name: dev
  haproxy:
    log: false
```

### compression

Turn on gzip compression.

If no [`compress_types`](#compress_types) are defined all responses are
compressed.

```yaml
environments:
- name: dev
  haproxy:
    compression: true
```

### compress_types

List of MIME types to compress.

Any value here implies [`compression: true`](#compression)

```yaml
environments:
- name: dev
  haproxy:
    compress_types: text/plain text/html application/json
```

### timeout

Set HAProxy timeouts.

**Warning: misconfigured timeouts can have an adverse effect on your service and
may make you more vulnurable to slowloris and DoS attacks. Change these at your
own risk**

The values in the following example are also the defaults if unset:

```yaml
environments:
- name: dev
  haproxy:
    timeout:
      client: 7s
      server: 7s
      tunnel: 7s
      http_request: 4s
      http_keep_alive: 60s
```

Durations are parsed by https://golang.org/pkg/time/#ParseDuration

`client` must equal `server`

`server` equals `client` if `server` is left unset

`tunnel` equals `client` if `tunnel` is left unset

Disable any timeout by setting the value to `0`

Refer to the [HAProxy docs](https://cbonte.github.io/haproxy-dconv/1.5/configuration.html#4.2-timeout%20client)
for what these timeouts mean

### maxconn

[`maxconn`](https://cbonte.github.io/haproxy-dconv/1.5/configuration.html#4-maxconn)
docs. This sets `maxconn` in the global and defaults sections.

Default: `200000`

### ssl

SSL support via HAProxy

### cert_directory

Provide an alternative path to the certs directory.

Default: `/etc/ssl/certs/`

### pem

Provide a path to an executable and any arguments to it. A PEM file should be
output on stdout. It's typically the private key concated with the issued cert.
[An example can be found here](https://www.digitalocean.com/community/tutorials/how-to-implement-ssl-termination-with-haproxy-on-ubuntu-14-04)

In this example porter is running on a build box with the cert on the filesystem
so it just `cat`s it.

**NOTE**: this is _not_ a suggestion of how to securely manage certs - only to
show how it works

```yaml
environments:
- name: dev
  haproxy:
    ssl:
      pem:
        secrets_exec_name: /bin/cat
        secrets_exec_args:
        - path/to/yourdomain_com.pem
```

### https_only

If true remove the HTTP port in security groups and remove the provisioned ELB's
HTTP listener.


CHANGELOG.md
`porter` is [semantically versioned](http://semver.org/spec/v2.0.0.html)

### v5.0.0

- build porter with Go 1.9.2
- default to `--pids-limit=4096`

### v4.9.0

- HAProxy `maxconn` is configurable
- fixed issue where `maxconn` wasn't set on the frontend

### v4.8.3

- enabling fix for volume mounts on SELinux by setting the environment variable `VOLUME_FLAG`

### v4.8.2

- allow selinux hosts to share the mounted volumes with the containers.
- allow ap-south-1 region

### v4.8.1

- replace deprecated sysctl setting

### v4.8.0

- HAProxy `timeout client` is configurable
- HAProxy `timeout server` is configurable
- HAProxy `timeout tunnel` is configurable
- HAProxy `timeout http-request` is configurable
- HAProxy `timeout http-keep-alive` is configurable

### v4.7.0

- build porter with Go 1.8.1
- optional ELB

### v4.6.0

- host-level SSL support

### v4.5.0

- added opt-in HAProxy compression
- added configurable list of MIME types to compress
- HAProxy logs can be turned off

### v4.4.0

- disabled userland proxy
- tuned network buffers

### v4.3.0

- added `c4.*`, `r4.*`, and `x1.*` instance types
- updated `m4.*` and `t2.*` instance types
- removed `g2.*`, `i2.*`, and `d2.*` instance types

### v4.2.0

- HAProxy stats endpoint auth is now randomized
- re-enabled keep-alive between HAProxy and containers
- build porter with Go 1.8
- added STANDARD_IA to secrets and CFN template uploads

### v4.1.1

- ASG size matching only occurs when `hot_swap: true`

### v4.1.0

- configurable instance count per region
- extended infrastructure ttl to a week

### v4.0.1

- fix type assertion for sg-ids that are statically defined

### v4.0.0

- failed stacks now delete instead of rollback
- lock down ASG egress traffic to allow by default NTP, DNS, HTTP, and HTTPS
- configurable haproxy header capture for logging

### v3.1.1

- service payloads were not encrypted as the docs said they were

### v3.1.0

- increased devicemapper base size to 50GB

### v3.0.5

- add `autowire_security_groups` so security group management can be turned off

### v3.0.4

- fixed issue with region-concurrent cleanup of service payload
- fixed possible issues with false-positive command success

### v3.0.3

- add `CREATE_IN_PROGRESS` to list of statuses that ignore ASG size
- add `DELETE_IN_PROGRESS` to list of statuses that ignore ASG size
- add `ROLLBACK_IN_PROGRESS` to list of statuses that ignore ASG size

### v3.0.2

- additional UPDATE steady states allow ASG matching
- any UPDATE in progress state causes hot swap to fail
- build porter with Go 1.7.3
- match currently promoted stack's ASG size for provisioning and hot swap

### v3.0.1

- allow 10 mins for service payload download+install during hot swap
- check for egress rules before writing `SecurityGroupEgress`

### v3.0.0

- updated to Amazon Linux 2016.09
- use Standard - Infrequent Access for service payload
- hot swap code on existing infrastructure
- kernel tuning allowing more concurrent connections
- added pre and post hotswap hooks
- fixed v2.4.3 issue that could create false-positives in `porter build` steps
- `net.ipv4.netfilter.ip_conntrack_tcp_timeout_time_wait = 1`
- added `cloudformation:DescribeStackResources` to deployment policy
- added `elasticloadbalancing:DescribeTags` to deployment policy
- added `sqs:CreateQueue` to deployment policy
- added `sqs:DeleteQueue` to deployment policy
- added `sqs:GetQueueAttributes` to deployment policy
- added `sqs:GetQueueUrl` to deployment policy
- added `sqs:ReceiveMessage` to deployment policy
- added `sqs:SendMessage` to ASG inline policy

### v2.4.3

- reject config files with `run_condition` set in a pre hook
- run post hooks with `run_condition` set to `fail` when a pre hook fails

### v2.4.2

- fix missing or incomplete hook logs

### v2.4.1

- gather hook log output by hook since they run concurrently

### v2.4.0

- added retries to instance autoregistration
- gather hook log output by region since they run concurrently
- log colorization is off by default
- run hooks concurrently across regions
- hook `run_condition`

### v2.3.0

- more resiliency for service payload downloads
- switch to sha-256 and validate service payload integrity
- extend container secret management to the host with `porter_get_secrets`
- fix support for running arbitrary user defined hooks

### v2.2.0

- run docker with `--security-opt=no-new-privileges`
- support docker registries as an alternative to S3
- support auto scaling group egress whitelist
- deprecated `dst_env_file`
- added `sse_kms_key_id` for optional SSE-KMS on all porter uploads

### v2.1.2

- increase logrotate size from 10M to 100M

### v2.1.1

- fix ec2-bootstrap hook clone for multi-region deployment
- configurable `-x` in `/var/log/cloud-init-output.log`
- service payload path is relative to support non-root volume

### v2.1.0

- `topology: worker` now supported
- configurable `read_only: false` to disable `docker run --read-only`

### v2.0.0

- improved secrets handling in transit
- enabled pluggable secrets provider
- locked down CloudFormation and S3 API call scopes to the resources needed
- service payload for S3 is now `{service name}/{environment}/{short sha}`
- add LOG_DEBUG environment variable for debug logging
- updated Amazon Linux AMI to 2016.03
- updated Docker to 1.11.2

conf/conf.go
/*
 * (c) 2016-2017 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

//
// Changing configuration fields or semantics?
// Make sure to update commands/help/files/config.go as well
//
package conf

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"regexp"
	"time"

	"github.com/adobe-platform/porter/constants"
	"github.com/adobe-platform/porter/stdin"
	"gopkg.in/inconshreveable/log15.v2"
	yaml "gopkg.in/yaml.v2"
)

const (
	Topology_Inet   = "inet"
	Topology_Worker = "worker"
	Topology_Cron   = "cron"
)

// NOTE: It's important to keep a reserved character so that if any of these
// things need to be concatenated we can split on the reserved character to
// get the original values unambiguously. That character is - so be very careful
// where it's allowed. Currently service names allow - in them meaning they can
// only be used on the beginning or end of a string
var (
	serviceNameRegex     = regexp.MustCompile(`^[a-zA-Z][-a-zA-Z0-9]*$`)
	roleARNRegex         = regexp.MustCompile(`^arn:aws:iam::\d+:role`)
	environmentNameRegex = regexp.MustCompile(`^[0-9a-zA-Z]+$`)
	healthMethodRegex    = regexp.MustCompile(`^GET$`)
	porterVersionRegex   = regexp.MustCompile(`^v\d+\.\d+\.\d+$`)
	vpcIdRegex           = regexp.MustCompile(`^vpc-(\d|\w){8}$`)
	subnetIdRegex        = regexp.MustCompile(`^subnet-(\d|\w){8}$`)
	commonNameRegex      = regexp.MustCompile(`^(\w+\.)?[a-z]+$`)

	// https://github.com/docker/docker/blob/v1.11.2/utils/names.go#L6
	// minus '-' which is reserved
	containerNameRegex = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9_.]+$`)
)

type (
	// Config supports multi-container deployments
	Config struct {
		ServiceName    string `yaml:"service_name"`
		ServiceVersion string
		PorterVersion  string            `yaml:"porter_version"`
		Environments   []*Environment    `yaml:"environments"`
		Slack          Slack             `yaml:"slack"`
		Hooks          map[string][]Hook `yaml:"hooks"`

		HAProxyStatsUsername string
		HAProxyStatsPassword string
	}

	Container struct {
		Name            string `yaml:"name"`
		OriginalName    string
		Topology        string       `yaml:"topology"`
		InetPort        int          `yaml:"inet_port"`
		Uid             *int         `yaml:"uid"`
		ReadOnly        *bool        `yaml:"read_only"`
		Dockerfile      string       `yaml:"dockerfile"`
		DockerfileBuild string       `yaml:"dockerfile_build"`
		HealthCheck     *HealthCheck `yaml:"health_check"`
		SrcEnvFile      *SrcEnvFile  `yaml:"src_env_file"`
		PidsLimit       int          `yaml:"pids_limit"`
	}

	SrcEnvFile struct {
		S3Key    string   `yaml:"s3_key"`
		S3Bucket string   `yaml:"s3_bucket"`
		S3Region string   `yaml:"s3_region"`
		ExecName string   `yaml:"exec_name"`
		ExecArgs []string `yaml:"exec_args"`
	}

	HealthCheck struct {
		Method string `yaml:"method"`
		Path   string `yaml:"path"`
	}

	Environment struct {
		Name                string           `yaml:"name"`
		StackDefinitionPath string           `yaml:"stack_definition_path"`
		RoleARN             string           `yaml:"role_arn"`
		Hotswap             bool             `yaml:"hot_swap"`
		InstanceCount       uint             `yaml:"instance_count"`
		InstanceType        string           `yaml:"instance_type"`
		BlackoutWindows     []BlackoutWindow `yaml:"blackout_windows"`
		Regions             []*Region        `yaml:"regions"`

		HAProxy HAProxy `yaml:"haproxy"`

		// From the client's perspective this relates to SG creation and ELB
		// inspection that allows the 2 ELBs to communicate with EC2 instances.
		// From porter's perspective this is just a signal to create them so
		// further transformations can happen
		CreateSecurityGroups *bool `yaml:"autowire_security_groups"`
	}

	HAProxy struct {
		// Capture headers for logging
		ReqHeaderCaptures []HeaderCapture `yaml:"request_header_captures"`
		ResHeaderCaptures []HeaderCapture `yaml:"response_header_captures"`
		Log               *bool           `yaml:"log"`
		Compression       bool            `yaml:"compression"`
		CompressTypes     []string        `yaml:"compress_types"`
		SSL               SSL             `yaml:"ssl"`
		Timeout           Timeout         `yaml:"timeout"`
		MaxConn           uint64          `yaml:"maxconn"`
	}

	Timeout struct {
		Client         *string `yaml:"client"`
		Client_        time.Duration
		Server         *string `yaml:"server"`
		Server_        time.Duration
		Tunnel         *string `yaml:"tunnel"`
		Tunnel_        time.Duration
		HttpRequest    *string `yaml:"http_request"`
		HttpRequest_   time.Duration
		HttpKeepAlive  *string `yaml:"http_keep_alive"`
		HttpKeepAlive_ time.Duration
	}

	SSL struct {
		CertDirectory  string `yaml:"cert_directory"`
		CertPath       string
		HTTPS_Only     bool `yaml:"https_only"`
		HTTPS_Redirect bool `yaml:"https_redirect"`

		Pem *Pem `yaml:"pem"`
	}

	Pem struct {
		SecretsExecName string   `yaml:"secrets_exec_name"`
		SecretsExecArgs []string `yaml:"secrets_exec_args"`
	}

	HeaderCapture struct {
		Header string `yaml:"header"`
		Length int    `yaml:"length"`
	}

	BlackoutWindow struct {
		StartTime string `yaml:"start_time"`
		EndTime   string `yaml:"end_time"`
	}

	Region struct {
		Name                string             `yaml:"name"`
		StackDefinitionPath string             `yaml:"stack_definition_path"`
		VpcId               string             `yaml:"vpc_id"`
		AZs                 []AvailabilityZone `yaml:"azs"`
		ELBs                []*ELB             `yaml:"elbs"`
		ELB                 string             `yaml:"elb"`
		RoleARN             string             `yaml:"role_arn"`
		AutoScalingGroup    AutoScalingGroup   `yaml:"auto_scaling_group"`
		SSLCertARN          string             `yaml:"ssl_cert_arn"`
		HostedZoneName      string             `yaml:"hosted_zone_name"`
		KeyPairName         string             `yaml:"key_pair_name"`
		S3Bucket            string             `yaml:"s3_bucket"`
		SSEKMSKeyId         *string            `yaml:"sse_kms_key_id"`
		Containers          []*Container       `yaml:"containers"`
		InstanceCount       uint               `yaml:"instance_count"`
	}

	AutoScalingGroup struct {
		SecurityGroupEgress []SecurityGroupEgress `yaml:"security_group_egress"`
		SecretsExecName     string                `yaml:"secrets_exec_name"`
		SecretsExecArgs     []string              `yaml:"secrets_exec_args"`
	}

	SecurityGroupEgress struct {
		CidrIp                     string `yaml:"cidr_ip" json:"CidrIp,omitempty"`
		FromPort                   int    `yaml:"from_port" json:"FromPort"`
		IpProtocol                 string `yaml:"ip_protocol" json:"IpProtocol,omitempty"`
		DestinationSecurityGroupId string `yaml:"destination_security_group_id" json:"DestinationSecurityGroupId,omitempty"`
		ToPort                     int    `yaml:"to_port" json:"ToPort"`
	}

	AvailabilityZone struct {
		Name     string `yaml:"name"`
		SubnetID string `yaml:"subnet_id"`
	}

	Hook struct {
		Repo         string            `yaml:"repo"`
		Ref          string            `yaml:"ref"`
		Dockerfile   string            `yaml:"dockerfile"`
		Environment  map[string]string `yaml:"environment"`
		Concurrent   bool              `yaml:"concurrent"`
		RunCondition string            `yaml:"run_condition"`
	}

	Slack struct {
		PackSuccessHook      string `yaml:"pack_success_webhook_url"`
		PackFailureHook      string `yaml:"pack_failure_webhook_url"`
		ProvisionSuccessHook string `yaml:"provision_success_webhook_url"`
		ProvisionFailureHook string `yaml:"provision_failure_webhook_url"`
		PromoteSuccessHook   string `yaml:"promote_success_webhook_url"`
		PromoteFailureHook   string `yaml:"promote_failure_webhook_url"`
	}

	ELB struct {
		ELBTag string `yaml:"tag"`
		Name   string `yaml:"name"`
	}
)

func (recv HAProxy) UsingSSL() bool {
	return recv.SSL.Pem != nil
}

func (recv *Config) GetEnvironment(envName string) (*Environment, error) {
	for _, env := range recv.Environments {
		if env.Name == envName {
			return env, nil
		}
	}
	return nil, fmt.Errorf("Environment %s doesn't exist in the config", envName)
}

// Convention over configuration
func (recv *Config) SetDefaults() {

	for _, hooks := range recv.Hooks {

		for i := 0; i < len(hooks); i++ {

			if hooks[i].RunCondition == "" {
				hooks[i].RunCondition = constants.HRC_Pass
			}
		}
	}

	for _, env := range recv.Environments {
		if env.InstanceCount == 0 {
			env.InstanceCount = 1
		}

		if env.InstanceType == "" {
			env.InstanceType = "m3.medium"
		}

		if len(env.HAProxy.CompressTypes) > 0 {
			env.HAProxy.Compression = true
		}

		if env.HAProxy.SSL.CertDirectory == "" {
			env.HAProxy.SSL.CertDirectory = "/etc/ssl/certs/"
		}

		if env.HAProxy.MaxConn == 0 {
			env.HAProxy.MaxConn = 200000
		}

		if env.HAProxy.Timeout.Client == nil || *env.HAProxy.Timeout.Client == "" {
			env.HAProxy.Timeout.Client = new(string)
			*env.HAProxy.Timeout.Client = "7s"
		}

		if env.HAProxy.Timeout.Server == nil || *env.HAProxy.Timeout.Server == "" {
			env.HAProxy.Timeout.Server = env.HAProxy.Timeout.Client
		}

		if env.HAProxy.Timeout.Tunnel == nil || *env.HAProxy.Timeout.Tunnel == "" {
			env.HAProxy.Timeout.Tunnel = env.HAProxy.Timeout.Client
		}

		if env.HAProxy.Timeout.HttpRequest == nil || *env.HAProxy.Timeout.HttpRequest == "" {
			env.HAProxy.Timeout.HttpRequest = new(string)
			*env.HAProxy.Timeout.HttpRequest = "4s"
		}

		if env.HAProxy.Timeout.HttpKeepAlive == nil || *env.HAProxy.Timeout.HttpKeepAlive == "" {
			env.HAProxy.Timeout.HttpKeepAlive = new(string)
			*env.HAProxy.Timeout.HttpKeepAlive = "60s"
		}

		env.HAProxy.SSL.CertPath = filepath.Join(env.HAProxy.SSL.CertDirectory, "porter.pem")

		// this is only for porterd which isn't currently informed of HTTPS_Only
		// porterd initially connects to http but will follow redirects
		if env.HAProxy.SSL.HTTPS_Only {
			env.HAProxy.SSL.HTTPS_Redirect = true
		}

		for _, region := range env.Regions {

			if len(region.AutoScalingGroup.SecurityGroupEgress) == 0 {
				region.AutoScalingGroup.SecurityGroupEgress = []SecurityGroupEgress{
					{ // DNS
						CidrIp:     "0.0.0.0/0",
						IpProtocol: "udp",
						FromPort:   53,
						ToPort:     53,
					},
					{ // NTP
						CidrIp:     "0.0.0.0/0",
						IpProtocol: "udp",
						FromPort:   123,
						ToPort:     123,
					},
					{ // HTTP
						CidrIp:     "0.0.0.0/0",
						IpProtocol: "tcp",
						FromPort:   80,
						ToPort:     80,
					},
					{ // HTTPS
						CidrIp:     "0.0.0.0/0",
						IpProtocol: "tcp",
						FromPort:   443,
						ToPort:     443,
					},
				}
			}

			// for the most part legacy behavior of looping through the ELB list
			// is still supported. copy newer-style config into older-style
			if region.ELB != "" {
				if region.ELBs == nil {
					region.ELBs = make([]*ELB, 0)
				}
				region.ELBs = append(region.ELBs, &ELB{
					Name: region.ELB,
				})
			}

			// support legacy behavior but ensure region.HasELB() works as
			// intended
			if region.ELB == "" && len(region.ELBs) > 0 {
				region.ELB = region.ELBs[0].Name
			}

			if len(region.Containers) == 0 {
				defaultContainer := &Container{}
				region.Containers = append(region.Containers, defaultContainer)
			}

			for _, container := range region.Containers {

				if container.Dockerfile == "" {
					container.Dockerfile = "Dockerfile"
				}
				if container.DockerfileBuild == "" {
					container.DockerfileBuild = "Dockerfile.build"
				}

				if container.PidsLimit == 0 {
					container.PidsLimit = 4096
				}

				if container.Topology == Topology_Inet {

					if container.HealthCheck == nil {
						container.HealthCheck = &HealthCheck{}
					}
					if container.HealthCheck.Method == "" {
						container.HealthCheck.Method = "GET"
					}
					if container.HealthCheck.Path == "" {
						container.HealthCheck.Path = "/health"
					}
				}
			}

			if len(region.Containers) == 1 {

				if region.Containers[0].Topology == "" {
					region.Containers[0].Topology = Topology_Inet
				}

				if region.Containers[0].Name == "" {
					region.Containers[0].Name = "primary"
				}
			}
		}
	}
}

func (recv *Config) Print() {
	fmt.Println("service_name", recv.ServiceName)
	fmt.Println("porter_version", recv.PorterVersion)

	fmt.Println(".Hooks")
	for hookName, hookVal := range recv.Hooks {
		printHooks(hookName, hookVal)
	}

	fmt.Println(".Environments")
	for _, environment := range recv.Environments {
		fmt.Println("- .Name", environment.Name)
		fmt.Println("  .StackDefinitionPath", environment.StackDefinitionPath)
		fmt.Println("  .RoleARN", environment.RoleARN)
		fmt.Println("  .InstanceCount", environment.InstanceCount)
		fmt.Println("  .InstanceType", environment.InstanceType)

		fmt.Println("  .Regions")
		for _, region := range environment.Regions {
			fmt.Println("  - .Name", region.Name)
			fmt.Println("    .VpcId", region.VpcId)
			fmt.Println("    .RoleARN", region.RoleARN)
			fmt.Println("    .KeyPairName", region.KeyPairName)
			fmt.Println("    .S3Bucket", region.S3Bucket)

			fmt.Println("      .AZs")
			for _, az := range region.AZs {
				fmt.Println("      - .Name", az.Name)
				fmt.Println("        .SubnetID", az.SubnetID)
			}

			fmt.Println("      .ELB", region.ELB)

			fmt.Println("      .ELBs")
			for _, elb := range region.ELBs {
				fmt.Println("      - .Tag", elb.ELBTag)
				fmt.Println("        .Name", elb.Name)
			}

			fmt.Println("      .Containers")
			for _, container := range region.Containers {
				fmt.Println("      - .Name", container.Name)
				fmt.Println("        .InetPort", container.InetPort)

				if container.Topology == Topology_Inet {
					fmt.Println("        .HealthCheck.Method", container.HealthCheck.Method)
					fmt.Println("        .HealthCheck.Path", container.HealthCheck.Path)
				}

				if container.SrcEnvFile == nil {
					fmt.Println("        .SrcEnvFile nil")
				} else {
					fmt.Println("        .SrcEnvFile.S3Key", container.SrcEnvFile.S3Key)
					fmt.Println("        .SrcEnvFile.S3Bucket", container.SrcEnvFile.S3Bucket)
					fmt.Println("        .SrcEnvFile.S3Region", container.SrcEnvFile.S3Region)
					fmt.Println("        .SrcEnvFile.ExecName", container.SrcEnvFile.ExecName)
					fmt.Println("        .SrcEnvFile.ExecArgs", container.SrcEnvFile.ExecArgs)
				}
			}
		}
	}
}

func printHooks(name string, hooks []Hook) {
	fmt.Println("  " + name)
	for _, hook := range hooks {
		fmt.Println("  - .Repo", hook.Repo)
		fmt.Println("    .Ref", hook.Ref)
		fmt.Println("    .Dockerfile", hook.Dockerfile)
		fmt.Println("    .Environment")
		if hook.Environment != nil {
			for envKey, envValue := range hook.Environment {
				if envValue == "" {
					envValue = os.Getenv(envKey)
				}
				fmt.Println("      " + envKey + "=" + envValue)
			}
		}
	}
}

func GetStdinConfig(log log15.Logger) (config *Config, success bool) {

	configBytes, err := stdin.GetBytes()
	if err != nil {
		log.Error("Failed to read from stdin", "Error", err)
		return
	}

	config, success = parseConfig(log, configBytes)
	return
}

func GetConfig(log log15.Logger, validate bool) (*Config, bool) {

	return getConfigFile(log, constants.ConfigPath, true, validate)
}

func GetAlteredConfig(log log15.Logger) (*Config, bool) {

	return getConfigFile(log, constants.AlteredConfigPath, false, false)
}

func GetHostConfig(log log15.Logger) (*Config, bool) {

	return getConfigFile(log, os.Getenv(constants.EnvConfigPath), false, false)
}

func getConfigFile(log log15.Logger, filePath string, setDefaults, validate bool) (config *Config, success bool) {
	var parseConfigSuccess bool

	file, err := os.Open(filePath)
	if err != nil {
		log.Error("os.Open", "FilePath", filePath, "Error", err)
		return
	}
	defer file.Close()

	configBytes, err := ioutil.ReadAll(file)
	if err != nil {
		log.Error("ioutil.ReadAll", "FilePath", filePath, "Error", err)
		return
	}

	config, parseConfigSuccess = parseConfig(log, configBytes)
	if !parseConfigSuccess {
		return
	}

	if setDefaults {

		config.SetDefaults()
	}

	if validate {

		err = config.Validate()
		if err != nil {
			log.Error("Config validation", "Error", err)
			return
		}
	}

	success = true
	return
}

func parseConfig(log log15.Logger, configBytes []byte) (config *Config, success bool) {
	config = &Config{}

	err := yaml.Unmarshal(configBytes, config)
	if err != nil {
		log.Error("Failed to decode config", "Error", err)
		return
	}

	success = true
	return
}
